### DISCUSSION TITLE
improve training and inference

### DISCUSSION THREAD

**Discussion by researcherofreality** [talmolab/sleap#2214] [Help!]
Hi,

First off, great software! It's been fun getting into.

I am trying to get sleap running on my experiment, where we have a single animal in a large arena. I've been looking through the issues and discussions to figure stuff out, but it's not really working, so I thought opening a discussion was worth it.

The setup is the following:
![image](https://github.com/user-attachments/assets/fef8ff18-a47e-44a3-a05c-a8135b05ec40)
![image](https://github.com/user-attachments/assets/bed646a3-0d9a-4499-afd1-774b33c31c54)

The whole video has a resolution of 2048x2440, at 40fps, with a length of ~8hrs. So far I have trained and predicted a few times, and totalled to 805 frames.

I'm using a top-down model, with quite good results. However, I'm having trouble with getting nice centered_instance results. The centroid seems to work just fine and quickly converges to a loss of 1e-5. However, the centered_instance seems to not capture the keypoints as nicely as I would like it to.

I've tried a few configs, mainly playing around with the input scaling, crop size, and model > max stride. I've tried max strides leading to receptive fields which are the size of the animal, as well as smaller ones. Changes I made and fixed throughout both centroid and centered_instance are: set max instance to 1, rotation to [-180, 180], and brightness [-5,5]. The latest run I did had the following config:

![image](https://github.com/user-attachments/assets/12ac4866-9df1-4e9b-b398-b9d4cb7e2967)
![image](https://github.com/user-attachments/assets/3dd73f79-7b8e-45b8-91e3-bdf0cb3a8725)
![image](https://github.com/user-attachments/assets/a0a17a95-9b75-425e-ac26-47d3f8b64f91)

I re-ran the above model with "Resume Training" but turning on "Online Mining", since that should resolve some of the harder to predict nodes, to little avail. Most prediction errors that occur I have adjusted before in the train, correct & re-train loop.

 Do you have any pointers as to how I could improve the centered_instance predictions? 

Another question is regarding the training setup time: it takes roughly 5 minutes to set up the training. Quicker would be nice for faster iterations; it's a lot of video to parse, but I would assume there's a way to speed this up. If there is, please let me know :) 
`INFO:sleap.nn.training:Setting up visualization...
INFO:sleap.nn.training:Finished trainer set up. [20.9s]
INFO:sleap.nn.training:Creating tf.data.Datasets for training data generation...
INFO:sleap.nn.training:Finished creating training datasets. [580.3s]
INFO:sleap.nn.training:Starting training loop...`

Also, the predictions which run afterwards for the metrics seem to be very slow, i.e. 1/2 fps, even though I am definitely using the GPU. When I run the below inference command, it goes at ~60fps. Any ideas? 

Then a question I had regarding inference. I am using the following command 
```
sleap-track "/path/to/video.mp4" 
-m "/path/to/centroid_dir" -m "/path/to/centered_instance_dir" 
--frames 18500-34000 -o "./sleap_output/output.slp" 
--open-in-gui  
--max_instances 1 
--tracking.tracker simple 
--tracking.similarity instance 
```

frames and open-in-gui is just to do a quick test, but more interestingly
1. I added the max_instances 1 to avoid having multiple instances & tracks, hopefully helping the model to infer and avoid fixing the tracks afterwards.
2. I added the simple tracker with similarity instance option to hopefully get more consistency in the inferred keypoints. Sometimes the animal moves very very slightly but a keypoint or two gets lost on the neural recording tether. I picked simple because I was observing memory leakage/bloating when using others, similar to  https://github.com/talmolab/sleap/issues/2081 

My question is, even when using the tracker some keypoints seem to jump on other objects, even when I explicitly fixed those in previous prediction&re-labeling. My assumption based on https://sleap.ai/develop/guides/proofreading.html#tracking-method-details is that using the tracker with instance similarity option should avoid this.: _" “instance” measures similarity by looking at the distances between corresponding nodes in the instances, normalized by the number of valid nodes in the candidate instance. "_
Should this work? Or should I try something else? Are there any other considerations you would have given the setup? My system and sleap config below.

Thank you!

```
memory     256GiB System memory
processor  AMD Ryzen Threadripper PRO 5975WX 32-Cores

 nvidia-smi
Fri Jun 27 16:11:39 2025
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 560.35.02              Driver Version: 560.94         CUDA Version: 12.6     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  NVIDIA GeForce RTX 4070 ...    On  |   00000000:41:00.0  On |                  N/A |
| 38%   68C    P2            190W /  220W |   11338MiB /  12282MiB |     86%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+
                                                                                 
+-----------------------------------------------------------------------------------------+
| Processes:                                                                              |
|  GPU   GI   CI        PID   Type   Process name                              GPU Memory |
|        ID   ID                                                               Usage      |
|=========================================================================================|
|    0   N/A  N/A       390      G   /Xwayland                                   N/A      |
|    0   N/A  N/A    641281    C+G   /sleap-label                                N/A      |
+-----------------------------------------------------------------------------------------+
```

Installed sleap using 
`conda create -y -n sleap -c conda-forge -c nvidia -c sleap/label/dev -c sleap -c anaconda sleap=1.4.1
`

switched to nvidia-cudnn-cu11==8.6.0.163 because of ncompatibility 

SLEAP: 1.4.1a2
TensorFlow: 2.7.0
Numpy: 1.21.6
Python: 3.7.12
OS: Linux-5.15.167.4-microsoft-standard-WSL2-x86_64-with-debian-trixie-sid

**Comment by gitttt-1234** [comment]
Hi @researcherofreality !

Could you please share a couple of screenshots where the switches happen?

Thanks, 

Divya

**Comment by researcherofreality** [comment]
Hi Divya, definitely, and thank you for helping out.

Below are some of the failure modes. Note that I have corrected quite a few of the same ones in the training&correction loop (800 labels total).

![image](https://github.com/user-attachments/assets/f82c3e50-7436-45b4-9c3d-deeebbb2bf06)
![image](https://github.com/user-attachments/assets/82ab63a6-af8b-4c6e-9ef1-ef02684c0412)

![image](https://github.com/user-attachments/assets/b1f51955-85e8-40da-8a21-67069f9285b8)
![image](https://github.com/user-attachments/assets/8d6a54d1-3726-4ab7-b6cf-a3a2a6069a91)

![image](https://github.com/user-attachments/assets/da21e3b0-256a-4a78-b576-fb8695b8092d)
![image](https://github.com/user-attachments/assets/50500baf-9cf0-4160-820a-b19a04503e06)

![image](https://github.com/user-attachments/assets/d936e59b-1cb9-4674-abab-2a1865d95052)
![image](https://github.com/user-attachments/assets/198be490-c409-41bc-9ede-5209cbd2eff7)

**Comment by talmo** [comment]
Hey @researcherofreality,

Regarding the tracking: you don't really have to do this if you just have a single animal, but the correct command would be:

```
sleap-track "/path/to/video.mp4" 
-m "/path/to/centroid_dir" -m "/path/to/centered_instance_dir" 
--frames 18500-34000 -o "./sleap_output/output.slp" 
--open-in-gui  
--max_instances 1 
--tracking.tracker simplemaxtracks
--tracking.max_tracks 1
```

If you keep it a bit simpler though, it won't assign tracks but that's ok for single animal data:

```
sleap-track "/path/to/video.mp4" 
-m "/path/to/centroid_dir" -m "/path/to/centered_instance_dir" 
--frames 18500-34000 -o "./sleap_output/output.slp" 
--max_instances 1
```

The [`sleap-convert`](https://sleap.ai/guides/cli.html#sleap-convert) and other commands should handle the single instance case without tracks assigned (but let us know if not).

As far as improving the keypoint localization performance: I would consider trying for a larger receptive field and maybe turning on online hard keypoint mining.

For centered instance:

**Online mining:** Checked
**Rotation min/max:** -180/180
**Max stride:** 32
**Filters:** 24
**Filters rate:** 1.5
**Anchor part:** tailstart1


Give that a go and let us know how it goes!

Cheers,

Talmo

**Reply by researcherofreality** [reply]
Thank you Talmo, I'll give it a shot now!

**Reply by researcherofreality** [reply]
Unfortunately didn't do the trick... Will play around with this a bit more - thanks!

**Comment by researcherofreality** [comment]
Another question I had was regarding inference on the cluster. I have consulted https://github.com/talmolab/sleap/discussions/833 but it's somewhat opaque as to what I should do specifically. 

The two main points seem to be 1. separating out CPU and GPU loads, and 2. alignment between GPUs expected by SLEAP and those on the cluster.

Regarding the first point: to me that seems to refer to first doing inference, and then the tracking, correct?

I submit the following bash script `cluster_run_day.sh`
```
CENTROID_MODEL="/path/to/best_model_250704_001327.centroid.n=1165"
INSTANCE_MODEL="/path/to/best_model_250704_011215.centered_instance.n=1165"
OUTPUT_DIR="./sleap_output"
INPUT_DIR="./data"

mkdir -p "$OUTPUT_DIR"

for mp4_file in "$INPUT_DIR"/compressed*.mp4; do
    base_name=$(basename "$mp4_file" .mp4)
    output_path="$OUTPUT_DIR/${base_name}.slp"
    analysis_path="$OUTPUT_DIR/${base_name}.analysis.h5"
    
    echo "Processing $mp4_file..."
    
    sleap-track \
        "$mp4_file" \
        -m "$CENTROID_MODEL" \
        -m "$INSTANCE_MODEL" \
        -o "$output_path" \
        --verbosity json \
        --batch_size 4 \
        --max_instances 1 
```

and then send to cluster using
`bsub -J 2025_06_09_npx06_day1 -o sleap_output/output.log -e sleap_output/error.log -q gpu_l4 -gpu "num=1" -R "rusage[mem=16000]" -n 1 bash -c "source ~/.bashrc && conda activate sleap && bash /path/to/cluster_run_day.sh"`

The L4 GPU has 15GB VRAM, and I also tried on A100s with 40GB VRAM, setting batch_size to 16/32. fps remained at 10... 

Question is, is this properly splitting the CPU and GPU load, or am I missing something?


Regarding the second point, the driver alignment part, I have the following:

```
(sleap) [Compute Node - user@h08u20]~>python - <<'EOF'
import tensorflow as tf
print("TF built for CUDA :", tf.sysconfig.get_build_info()["cuda_version"])
import ctypes, ctypes.util
rt = ctypes.CDLL(ctypes.util.find_library("cudart") or "libcudart.so")
ver = ctypes.c_int()
rt.cudaRuntimeGetVersion(ctypes.byref(ver))
print("Runtime CUDA vers:", f"{ver.value//1000}.{(ver.value%1000)//10}")
EOF
TF built for CUDA : 11.2
Runtime CUDA vers: 11.2
```

So if I understand correctly, it indeed has the correct drivers configured in the environment, which should in turn be used by tensorflow for speedy processing; nevertheless, inference remains at ~10fps.

Apologies for the lengthy post, I hope it's kinda clear - hope you have some pointers, and thank you very much for the assistance!

**Reply by researcherofreality** [reply]
Interestingly, locally (RTX4070 super) running the same script as above I have ~50 fps, which is already somewhat better, but not close to the 100s, and can't run multiple videos at once. Also, the output from the drivers check seems to be the same. Funky stuff.
